### PR TITLE
175566498 warm start qaoa

### PR DESCRIPTION
**Pull request by mstechly**

### CODE REVIEW

**Comment by Guoming-Wang** [comment]
Looks great!

**Review by Guoming-Wang — APPROVED** [review]

File: src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+    supports_parametrized_circuits = True
+    cost_hamiltonian = ansatz_property("cost_hamiltonian")
+    thetas = ansatz_property("thetas")
+
+    def __init__(
+        self,
+        number_of_layers: int,
+        cost_hamiltonian: Union[QubitOperator, IsingOperator],
+        thetas: np.ndarray,
```
Comment (simonwa7) [inline src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py:24]: `thetas` input is not described

File: src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    @property
+    def number_of_params(self) -> int:
+        """Returns number of parameters in the ansatz."""
+        return 2 * self.number_of_layers
+
+    @overrides
+    def _generate_circuit(self, params: Optional[np.ndarray] = None) -> Circuit:
+        """Returns a parametrizable circuit represention of the ansatz.
+        By convention the initial state is taken to be the |+..+> state and is
```
Comment (simonwa7) [inline src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py:54]: This is different than the initial state prepared in the code which I believe uses parametrized Ry's on each qubit

File: src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                "This method retuns a parametrizable circuit, params will be ignored."
+            )
+        circuit = Circuit()
+        qubits = [Qubit(qubit_index) for qubit_index in range(self.number_of_qubits)]
+        circuit.qubits = qubits
+
+        # Prepare initial state
+        circuit += create_layer_of_gates(self.number_of_qubits, "Ry", self._thetas)
+
+        # Add time evolution layers
```
Comment (simonwa7) [inline src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py:70]: comment seems incorrect

File: src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        if params is not None:
+            Warning(
+                "This method retuns a parametrizable circuit, params will be ignored."
+            )
+        circuit = Circuit()
+        qubits = [Qubit(qubit_index) for qubit_index in range(self.number_of_qubits)]
+        circuit.qubits = qubits
+
+        # Prepare initial state
+        circuit += create_layer_of_gates(self.number_of_qubits, "Ry", self._thetas)
```
Comment (simonwa7) [inline src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py:68]: Just commenting to resolve conflict in function header

File: src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        number_of_layers: int,
+        cost_hamiltonian: Union[QubitOperator, IsingOperator],
+        thetas: np.ndarray,
+    ):
+        """TODO
+
+        Args:
+            number_of_layers: number of layers of the ansatz. Also refered to as "p" in the paper.
+            cost_hamiltonian: Hamiltonian representing the cost function
+            mixer_hamiltonian: Mixer hamiltonian for the QAOA. If not provided, will default to basic operator consisting of single X terms.
```
Comment (simonwa7) [inline src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py:31]: mixer_hamiltonian is not an input

Comment (simonwa7) [reply src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py:31]: Additionally, should emphasize that `thetas` shouldn't be symbolic params

File: src/python/zquantum/qaoa/ansatzes/warm_start_ansatz_test.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        return np.array([0.5, 0.5])
+
+    @pytest.fixture
+    def ansatz(self, thetas):
+        cost_hamiltonian = QubitOperator((0, "Z")) + QubitOperator((1, "Z"))
+        return WarmStartQAOAAnsatz(
+            number_of_layers=1,
+            cost_hamiltonian=cost_hamiltonian,
+            thetas=thetas,
+        )
```
Comment (simonwa7) [inline src/python/zquantum/qaoa/ansatzes/warm_start_ansatz_test.py:24]: I think it's important to test the ansatz with at least 2 layers in addition to just the 1 layer ansatz

File: src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    cost_hamiltonian = ansatz_property("cost_hamiltonian")
+    thetas = ansatz_property("thetas")
+
+    def __init__(
+        self,
+        number_of_layers: int,
+        cost_hamiltonian: Union[QubitOperator, IsingOperator],
+        thetas: np.ndarray,
+    ):
+        """TODO
```
[the remark this answers is not in the document]
Comment (mstechly) [reply src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py]: Fixed in https://github.com/zapatacomputing/z-quantum-qaoa/pull/31/commits/0bdf13c508c63ae5f8f26ed819fb7b57b4ba9110

File: src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+        Args:
+            number_of_layers: number of layers of the ansatz. Also refered to as "p" in the paper.
```
[the remark this answers is not in the document]
Comment (mstechly) [reply src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py:32]: Fixed in https://github.com/zapatacomputing/z-quantum-qaoa/pull/31/commits/0bdf13c508c63ae5f8f26ed819fb7b57b4ba9110

File: src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py (hunk 1 above)
Comment (mstechly) [reply src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py:24]: Fixed in https://github.com/zapatacomputing/z-quantum-qaoa/pull/31/commits/0bdf13c508c63ae5f8f26ed819fb7b57b4ba9110

File: src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py (hunk 2 above)
Comment (mstechly) [reply src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py]: Fixed in https://github.com/zapatacomputing/z-quantum-qaoa/pull/31/commits/0bdf13c508c63ae5f8f26ed819fb7b57b4ba9110

File: src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py (hunk 3 above)
Comment (mstechly) [reply src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py]: Fixed in https://github.com/zapatacomputing/z-quantum-qaoa/pull/31/commits/0bdf13c508c63ae5f8f26ed819fb7b57b4ba9110

File: src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py (hunk 5 above)
Comment (mstechly) [reply src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py]: Fixed in https://github.com/zapatacomputing/z-quantum-qaoa/pull/31/commits/0bdf13c508c63ae5f8f26ed819fb7b57b4ba9110

File: src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py (hunk 4 above)
Comment (mstechly) [reply src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py:69]: Not sure what you mean?

**Comment by mstechly** [comment]
In https://github.com/zapatacomputing/z-quantum-qaoa/pull/31/commits/dec1970466d3d32d92abeb64e80b77dba340aa07 I moved `convert_relaxed_solution_to_angles` here, as it makes more sense to keep it close to WS-QAOA ansatz.

File: src/python/zquantum/qaoa/ansatzes/warm_start_ansatz_test.py (hunk 6 above)
Comment (mstechly) [reply src/python/zquantum/qaoa/ansatzes/warm_start_ansatz_test.py]: Here you are: https://github.com/zapatacomputing/z-quantum-qaoa/pull/31/commits/eb41b606cd948b9b4f8e664ca8165d7e33742fa7

File: steps/qaoa.py [hunk 9]
```diff
@@ -55,3 +59,10 @@ def get_maxcut_hamiltonian(graph, scaling=1.0, shifted=False):
         graph_object, scaling=scaling, shifted=shifted
     )
     save_qubit_operator(hamiltonian, "hamiltonian.json")
+
+
+def convert_relaxed_solution_to_angles(solution, epsilon=0.5, decimals=8):
+    solution = np.around(load_list(solution), decimals=decimals)
```
Comment (simonwa7) [inline steps/qaoa.py:65]: How come we're rounding these values?

Comment (simonwa7) [reply steps/qaoa.py:65]: Follow-up, is this the right place to round these numbers? Would it make more sense for them to be rounded inside of `convert_relaxed_solution_to_angles`?

**Review by simonwa7 — APPROVED** [review]

File: src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    thetas = ansatz_property("thetas")
+
+    def __init__(
+        self,
+        number_of_layers: int,
+        cost_hamiltonian: Union[QubitOperator, IsingOperator],
+        thetas: np.ndarray,
+    ):
+        """
+        This is implementationg of the warm-start QAOA Ansatz from https://arxiv.org/abs/2009.10095v3 .
```
Comment (dexter2206) [inline src/python/zquantum/qaoa/ansatzes/warm_start_ansatz.py:27]: Typo: implementationg -> implementation.

File: src/python/zquantum/qaoa/ansatzes/warm_start_ansatz_test.py [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+import numpy as np
+import sympy
+
+
+def create_betas(number_of_layers):
+    return list(sympy.symbols(f"beta_:{number_of_layers}"))
+
+
+def create_gammas(number_of_layers):
+    return list(sympy.symbols(f"gamma_:{number_of_layers}"))
```
Comment (dexter2206) [inline src/python/zquantum/qaoa/ansatzes/warm_start_ansatz_test.py:17]: Based on how those functions are used, it seems to me that conversion to list is redundant here.

File: steps/qaoa.py [hunk 12]
```diff
@@ -55,3 +59,9 @@ def get_maxcut_hamiltonian(graph, scaling=1.0, shifted=False):
         graph_object, scaling=scaling, shifted=shifted
     )
     save_qubit_operator(hamiltonian, "hamiltonian.json")
+
+
+def convert_relaxed_solution_to_angles(solution, epsilon=0.5):
+    solution = np.array(load_list(solution))
+    thetas = _convert_relaxed_solution_to_angles(solution, epsilon)
+    save_list(thetas.tolist(), "thetas.json")
```
Comment (dexter2206) [inline steps/qaoa.py:67]: No newline at the end of the file.

**Review by dexter2206 — APPROVED** [review]

**Pull request merged**